clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java

```diff
@@ -930,10 +940,15 @@ private boolean filterRecordByVersionSwapHighWatermarks(
     return false;
   }
 
+  protected Set<PubSubTopicPartition> getTopicAssignment() {
+    return Collections.synchronizedSet(pubSubConsumer.getAssignment());
```

Comment:
This would just return a synchronized set to the caller. IIUC we need to add sync on `pubSubConsumer` in places like switchToNewTopic to prevent concurrent exception.

Reply:
The exception thrown from earlier is because the assignment collection returned isn't a copy.  This change addresses that particular problem.

I'm not a big fan of adding synchronized blocks unless theres a strong logical need to.  My scan through of uses of pubsubconsumer seem to indicate that most of the calls are already protected by synchronized calls, and where it's not I don't see an obvious concurrency issue.  Are there calls in particular you are worried about?

Reply:
Per javadoc, even the set has Synchronized keyword on it, it still need be put a synchronzied keyword around the traverse. Is it what you want Zac? would a ConcurrentSkipListSet or concurrencyHashMap do a better job? 

Per the javadoc, even the set has Synchronized keyword on it, it still need be put a synchronzied keyword around the traverse. Is it what you want Zac? would a ConcurrentSkipListSet or concurrencyHashMap do a better job?

Reply:
Which is what the PR is already doing.

But more aside to the problem, even just calling getAssignment() on the pubsubconsumer can barf, so the underlying implementation of pubsubconsumer is entirey non thread safe.  So the PR has just been updated to synchronize all access instead.

Reply:
I went back to double check the places that call getTopicAssignment. And almost all places used synchronized except for one, line 953 of VeniceChangelogConsumerImpl.java, which simply traverses the set. So some thread's change on the thread may not be seen by whoever runs on line 953. Are we good with it?

Reply:
Yeah, this context is pretty safe.  Making the copy is probably cheaper.  However, we have to use synchronization because of the implementation of the KafkaPubSubConsumer.